vortex-array/src/arrays/chunked/compute/take.rs

```diff
@@ -1,44 +1,111 @@
 use vortex_buffer::BufferMut;
-use vortex_dtype::PType;
+use vortex_dtype::{DType, Nullability, PType};
 use vortex_error::VortexResult;
+use vortex_mask::Mask;
 
-use crate::arrays::ChunkedVTable;
 use crate::arrays::chunked::ChunkedArray;
+use crate::arrays::{ChunkedVTable, PrimitiveArray};
 use crate::compute::{TakeKernel, TakeKernelAdapter, cast, take};
+use crate::validity::Validity;
 use crate::{Array, ArrayRef, IntoArray, ToCanonical, register_kernel};
 
 impl TakeKernel for ChunkedVTable {
     fn take(&self, array: &ChunkedArray, indices: &dyn Array) -> VortexResult<ArrayRef> {
-        let indices = cast(indices, PType::U64.into())?.to_primitive()?;
-
-        // While the chunk idx remains the same, accumulate a list of chunk indices.
-        let mut chunks = Vec::new();
-        let mut indices_in_chunk = BufferMut::<u64>::empty();
-        let mut prev_chunk_idx = array
-            .find_chunk_idx(indices.as_slice::<u64>()[0].try_into()?)
-            .0;
-        for idx in indices.as_slice::<u64>() {
-            let idx = usize::try_from(*idx)?;
-            let (chunk_idx, idx_in_chunk) = array.find_chunk_idx(idx);
-
-            if chunk_idx != prev_chunk_idx {
-                // Start a new chunk
-                let indices_in_chunk_array = indices_in_chunk.clone().into_array();
-                chunks.push(take(array.chunk(prev_chunk_idx)?, &indices_in_chunk_array)?);
-                indices_in_chunk.clear();
-            }
-
-            indices_in_chunk.push(idx_in_chunk as u64);
-            prev_chunk_idx = chunk_idx;
+        let indices = cast(
+            indices,
+            &DType::Primitive(PType::U64, indices.dtype().nullability()),
+        )?
+        .to_primitive()?;
+
+        if indices.dtype().is_nullable() {
+            take_nullable(array, indices.as_slice::<u64>(), indices.validity_mask()?)
+        } else {
+            take_non_nullable(array, indices.as_slice::<u64>())
+        }
```

Comment:
I'm curious if there's a significant slowdown to adding a mask slice if the mask is alltrue.

Reply:
Hmm. Yep. I'll include and rename the non null arm

Reply:
as in - do we need to distinguish nullable and non nullable branches at all? In non nullable case the Mask is going to be AllTrue and we just pay a little bit of extra overhead of calling slice to resize it

Reply:
There is definitely some overhead, hard to know how much, since the fn is pretty small shall we keep it?

Reply:
I would keep it simple for now and refactor if necessary in the future

Reply:
Sure if you prefer